Add img_mgmt_get_next_boot_slot  and switch upload and erase logic to use it

de-nordic opened the pull request:
The PR contains series of commits that:
1) Provide img_mgmt_get_next_boot_slot  and img_mgmt_get_opposite_slot functions allowing to  obtain information on next boot slot for given image and type of boot and opposite slot;
2) Switches various function to use the above function for detecting type of next boot;
3) Improves image erase logic;
4) Adds additional Kconfig options to allow better control on setting image flags and erase process;
5) Makes code more image number independent and reduces #ifdefs dedicated to supported number of images.

**Update 2023-08-17 15:00 UTC**
Applied requested changes.
**Update 2023-08-18 17:50 UTC**
Applied set of review comments and fixed problem with `CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY`. There is one additional commit now that reduces `img_mgmt_slot_to_image`.
**Update 2023-08-18 18:20 UTC**
Minor fix in commit message.
**Update 2023-08-24 10:24 UTC**
Review fixes and update from main.
**Update 2023-08-24 12:12 UTC**
Rebase onto main.
**Update 2023-08-24 14:45 UTC**
Added some initialization on output variable to make static analysis happy.

de-nordic committed:
mgmt/MCUmgr/grp/img: Add img_mgmt_get_next_boot_slot

The commit adds functions:
 img_mgmt_get_next_boot_slot
 img_mgmt_get_opposite_slot

to simplify obtaining information on next boot slot.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/MCUmgr/grp/img: Switch to using img_mgmt_get_next_boot_slot

The commit modifies image list command operations to use
img_mgmt_get_next_boot_slot instead of directly relying of
MCUboot flags.
The function is now used, also, by img_mgmt_slot_in_use to
figure out whether queried slot is in use.
The commit introduces two new Kconfig options
 MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY
 MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY
that allow users to enable confirming non-active images slots.
The MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY is y
by default to keep original behavior of logic that accidentally
allowed confirming secondary slot.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/MCUmgr/grp/img: Reduce logic in img_mgmt_active_slot

Make img_mgmt_active_slot independent from
CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER for MCUboot swap type
algorithms.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
doc/release/3.5: Add note on new MCUmgr Kconfig options

Note on addition of:
  CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY
  CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY
  CONFIG_MCUMGR_GRP_IMG_ALLOW_ERASE_PENDING

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/MCUmgr/grp/img: Improvement in image erase logic

The erase logic has been switched to using img_mgmt_get_opposite_slot
and the img_mgmt_get_other_slot has been removed.
The commit adds CONFIG_MCUMGR_GRP_IMG_ALLOW_ERASE_PENDING Kconfig
options, default set to n, that allows to make pending slot
erasable. The option only allows erase on pending slot that
is not revert slot.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/MCUmgr/grp/img: Simplify img_mgmt_get_unused_slot_area_id

Simplified and made CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER
independent.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:
```diff
@@ -61,6 +61,39 @@ config MCUMGR_GRP_IMG_VERBOSE_ERR
+
+config MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY
+	bool "Allow to confirm slots of non-active image"
+	default n
```
`default n` can go

de-nordic replied:
Done

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:
```diff
@@ -61,6 +61,39 @@ config MCUMGR_GRP_IMG_VERBOSE_ERR
+config MCUMGR_GRP_IMG_ALLOW_ERASE_PENDING
+	bool "Allow to erase pending slot"
+	help
+	  Allows to erase secondary slot set for test or confirmed; this allows
+	  to erase slot that has already been set for next boot but there has been
+	  no reset yet so it has not yet been swapped.
```
```suggestion
	  Allows erasing secondary slot which is marked for test or confirmed; this allows
	  erasing slots that have been set for next boot but the device has not
	  reset yet, so has not yet been swapped.
```
Also this refers to secondary but in the case of direct XIP it could be primary, maybe "opposite" slot would be useful, I think all Kconfig options have this issue.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/include/mgmt/mcumgr/grp/img_mgmt/img_mgmt_priv.h:
```diff
@@ -86,6 +86,51 @@ int img_mgmt_write_image_data(unsigned int offset, const void *data, unsigned in
+ */
+int img_mgmt_get_next_boot_slot(int image, int *type);
+
+#define NEXT_BOOT_TYPE_NORMAL	0
+#define NEXT_BOOT_TYPE_TEST	1
+#define NEXT_BOOT_TYPE_REVERT	2
```
Probably easier to make an enum and then instead of `int *type` have `enum x *type` so the documentation links to the type

de-nordic replied:
Done

nordicjm reviewed (commented):
minor nits, not fully checked code yet

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c:
```diff
@@ -141,9 +141,10 @@ static int img_mgmt_find_tlvs(int slot, size_t *start_off, size_t *end_off, uint
+	LOG_DBG("(%d)", image);
+#if CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER > 1
+	if (image >= 1) {
+		return (image << 1);
```
Not requesting a change but this will probably cause a problem in the future because it assumes that the primary slot in other images is the one executing, but if other images are using direct XIP then this is an incorrect assumption

de-nordic replied:
Yes, that is known problem. Without communicating with other app or pulling info from MCUboot, for example via retention, it is hard to tell what is actually active slot of another image as what is set to flash may not yet be actually executing.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +139,91 @@ img_mgmt_state_flags(int query_slot)
+int img_mgmt_get_opposite_slot(int slot)
+{
+	if (slot < 0 ||
+	    slot > (CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER) << 1) {
```
`>=`? If 2 images, slots are 0, 1, 2, 3 but 2 << 1 is 4

de-nordic replied:
Good catch.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +139,91 @@ img_mgmt_state_flags(int query_slot)
+
+	switch (state) {
+	case BOOT_SWAP_TYPE_NONE:
+			slot = image << 1;
```
wonky indent

de-nordic replied:
Done

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +139,91 @@ img_mgmt_state_flags(int query_slot)
+			const int active_slot = img_mgmt_active_slot(image);
+
+			slot = img_mgmt_get_opposite_slot(active_slot);
+			/* App has boot to be tested and has not yet been confirmed, which means
```
*Application has to boot to be tested

de-nordic replied:
Done

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +139,91 @@ img_mgmt_state_flags(int query_slot)
+
+			slot = img_mgmt_get_opposite_slot(active_slot);
+			/* App has boot to be tested and has not yet been confirmed, which means
+			 * that next boot will be revert to reported slot.
```
*that next boot the image will revert to the reported slot.

de-nordic replied:
Done

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +139,91 @@ img_mgmt_state_flags(int query_slot)
+	int rcs = img_mgmt_read_info(other_slot, &over, NULL, NULL);
+	int rca = img_mgmt_read_info(active_slot, &aver, NULL, NULL);
+
+	if (rcs == 0 && rca == 0 && img_mgmt_vercmp(&aver, &over) < 0) {
```
(likewise with this, I wonder how a direct XIP configured system would work on e.g. nRF53 where the secondary image would not be direct XIP as it would be loaded to the network core, maybe we need some new option in dts per partition which specifies what type of upgrade system it uses... seems like a messy idea)

de-nordic replied:
Currently we do not support DirectXIP with nRF53 as this would required mixed behaviour where app core image is DirectXIPed but the net core needs to be transferred to to the core, which means swapped.

I would actually prefer the nRF53 two core boots to desist using MCUboot for swapping net core image and instead drop that job on an application: this way MCUboot has to care about booting the app core image only and net core will be handled the same way whether swap or DirectXIP is used. Also, if net core image becomes part of app core image we have greater flexibility in flash partitioning, and less fragmentation, as we do not have to use special partitions for net core anymore.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -350,7 +533,11 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
 #if defined(CONFIG_MCUMGR_GRP_IMG_STATUS_HOOKS)
-	if (active) {
+	int32_t ret_rc;
+	uint16_t ret_group;
+
+	if (slot == active_slot && confirm) {
+		uint16_t ret_group;
```
Shadow variable

de-nordic replied:
Done

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +139,92 @@ img_mgmt_state_flags(int query_slot)
+			const int active_slot = img_mgmt_active_slot(image);
+
+			slot = img_mgmt_get_opposite_slot(active_slot);
+			/* Application has to boot to be testd and has not yet been confirmed,
```
*tested

de-nordic replied:
Done

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +139,92 @@ img_mgmt_state_flags(int query_slot)
+		break;
+	case BOOT_SWAP_TYPE_PERM:
+		/* Type is NEXT_BOOT_TYPE_NORMAL as it means that the returned slot
+		 * is boot slot is normal slot for all next boots.
```
Is this intentional "is boot slot is normal slot"?

de-nordic replied:
Better comment provided.

nordicjm reviewed (commented):
minor nits.

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/include/mgmt/mcumgr/grp/img_mgmt/img_mgmt_priv.h:
```diff
@@ -86,6 +86,59 @@ int img_mgmt_write_image_data(unsigned int offset, const void *data, unsigned in
+	/* Next boot will be revert to already confirmed slot; this
+	 * type of next boot means that active slot is not confirmed
+	 * yet as it has been marked for test in previous boot.
+	 */
```
Done

MarekPieta commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -392,35 +429,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
+	 * a bug in logic that always allowed to confirm secondary slot of any
+	 * image. Now the behaviour is controlled via Kconfig options.
+	 */
+	if ((!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY) ||
+	     !IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY)) &&
+	    confirm && image != img_mgmt_active_image() &&
+	    (!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY) ||
+	     slot == active_slot)) {
+		LOG_DBG("Not allowed to confirm non-active images");
+		return IMG_MGMT_RET_RC_INVALID_SLOT;
+	}
```
```suggestion
if (confirm) {
    if (slot == active_slot) {
        if (!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY) &&
            image != img_mgmt_active_image()) {
                return IMG_MGMT_RET_RC_INVALID_SLOT;
            }
    } else {
        if (!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY)) {
            return IMG_MGMT_RET_RC_INVALID_SLOT;
        }
    }
}
```

de-nordic replied:
Selected different approach.

de-nordic committed:
mgmt/MCUmgr/grp/img: Reduce logic in img_mgmt_slot_to_image

Makes img_mgmt_slot_to_image image number independent and moves
it to header file as static inline.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/include/mgmt/mcumgr/grp/img_mgmt/img_mgmt_priv.h:
```diff
@@ -86,6 +86,59 @@ int img_mgmt_write_image_data(unsigned int offset, const void *data, unsigned in
+ * boot on next reset
+ * @return -1 in case when next boot slot can not be established.
+ * @p type Will be set to next boot type. There are currently three types
+ * defined:
```
Done

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +134,95 @@ img_mgmt_state_flags(int query_slot)
+			const int active_slot = img_mgmt_active_slot(image);
+
+			slot = img_mgmt_get_opposite_slot(active_slot);
+			/* Application has to boot to be tested and has not yet been confirmed,
```
Done

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +134,95 @@ img_mgmt_state_flags(int query_slot)
+
+	switch (state) {
+	case BOOT_SWAP_TYPE_NONE:
+		slot = image << 1;
```
I have moved the code to use `img_mgmt_active_slot` but I do not know yet whether this will be used in DIrectXIP with revert.

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +134,95 @@ img_mgmt_state_flags(int query_slot)
+		 * Normal here means that it is ordinary boot and slot has not been marked
+		 * for revert or pending for test, and may change on reset.
+		 */
+		slot = (image << 1) + 1;
```
Fixed.

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +134,95 @@ img_mgmt_state_flags(int query_slot)
+		}
+		break;
+	default:
+		/* Should never, ever happen */
```
I want to leave this with LOG_DBG to know what has been reported here. We can change this in other PR if you are not satisfied.

MarekPieta replied:
I wanted to highlight it more, but I wound not insist. We can leave it as is.

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/zephyr_img_mgmt.c:
```diff
@@ -238,13 +238,15 @@ static int img_mgmt_get_unused_slot_area_id(int slot)
+	if (image == -1) {
+		image = 0;
+	}
+	slot = (image << 1) + 1;
```
Done

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -392,35 +429,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
-			return MGMT_ERR_EINVAL;
+	/* Can not change state when already set to test the non-active image. */
+	if (type == NEXT_BOOT_TYPE_TEST) {
+		return IMG_MGMT_RET_RC_IMAGE_ALREADY_PENDING;
```
Done

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +134,95 @@ img_mgmt_state_flags(int query_slot)
+/* Return number of slot that is considered candidate for next boot */
+int img_mgmt_get_opposite_slot(int slot)
+{
+	if (slot < 0 ||
```
Done

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -375,35 +412,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
+		return IMG_MGMT_RET_RC_INVALID_SLOT;
+	}
+
+	/* DO nothing when requested to test the non-active image already set for test. */
```
*Do

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c:
```diff
@@ -141,16 +141,21 @@ static int img_mgmt_find_tlvs(int slot, size_t *start_off, size_t *end_off, uint
 	if (FIXED_PARTITION_IS_RUNNING_APP_PARTITION(slot1_partition)) {
-		return 1;
+		slot = 1;
 	}
-	return 0;
+	slot = 0;
```
this will always return 0

MarekPieta replied:
else seems needed

de-nordic replied:
No, that is rebase error, the `slot = 0` should not be here at all.

nordicjm reviewed (changes requested):
only last one and typo are required changes, other is comment

MarekPieta commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -375,35 +412,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
+		if (slot == active_slot && active_slot != next_boot_slot) {
+			return IMG_MGMT_RET_RC_IMAGE_ALREADY_PENDING;
+		}
+		/* Allow selecting non-active slot for boot */
```
I wonder if we shouldn't disallow the following case too `!confirm && slot == next_boot_slot && active_slot != next_boot_slot` (to prevent trying to set test flag for the `next_boot_slot` when permanent swap is pending)

MarekPieta commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -233,8 +233,23 @@ img_mgmt_slot_in_use(int slot)
+	    (active_slot != nbs && type == NEXT_BOOT_TYPE_NORMAL)) {
+#if defined(CONFIG_MCUMGR_GRP_IMG_ALLOW_ERASE_PENDING)
+		LOG_DBG("(%d) Allowed erase pending", slot);
+		return 0;
```
`return (active_slot == slot)`
Otherwise you may allow to erase the currently running slot if the secondary slot marked as permanent swap.,

nvlsianpu replied:
second check will be more legible if current boot type would be used:
slot == nbs && curr_type == BOOT_TYPE_TEST

nvlsianpu commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/zephyr_img_mgmt.c:
ADD_SLOT_2_CONDITION and #define ADD_SLOT_3_CONDITION define can be removed

de-nordic replied:
Removed

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -375,35 +412,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
+	    (!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY) ||
+	     slot == active_slot)) {
+		LOG_DBG("Not allowed to confirm non-active images");
+		return IMG_MGMT_RET_RC_INVALID_SLOT;
```
So generally I should drop the entire check?

de-nordic replied:
Previously we did not allow to confirm non-running image.

de-nordic replied:
@nordicjm  Should I completely drop the `CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY` and `CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY` ?

nordicjm replied:
There doesn't look to be an hook in this code which allows the change to be declined, that would be required to drop this, in my mind this should be entirely user/application driven, if a user wants to allow or deny confirming slots for specific reasons, they should decide on those rules because they will probably be application and mcuboot operating mode specific

nordicjm replied:
Leave this in but update the error code, hook version will be a "version 2: in the future!" feature, probably when someone asks for it

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -133,6 +134,78 @@ img_mgmt_state_flags(int query_slot)
+	const int active_slot = img_mgmt_active_slot(image);
+	const int state = mcuboot_swap_type_multi(image);
+	/* All cases except BOOT_SWAP_TYPE_NONE return opposite slot */
+	int slot = img_mgmt_get_opposite_slot(active_slot);
```
Comment explains why thing happens.
Code does something because something else, but does not explain why.

MarekPieta replied:
The comment would be reflected by the code with the suggested change, but we can keep it as is.

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -375,35 +412,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
+
+	/* DO nothing when requested to test the non-active image already set for test. */
+	if (type == NEXT_BOOT_TYPE_TEST) {
+		return 0;
```
Adding test for `confirm`.

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -375,35 +412,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
 	if (rc != 0) {
 		/* Failed to set next slot for boot as desired */
-		if (active) {
+		if (slot == active_slot) {
```
Added

MarekPieta commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -375,39 +412,87 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
-	if (active && img_mgmt_state_any_pending()) {
+	if (type == NEXT_BOOT_TYPE_TEST) {
+		/* Do nothing when requested to test the slot already set for test. */
+		if (slot == next_boot_slot && confirm == false) {
```
`!confirm && slot == next_boot_slot` (for consistency with existing code)

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -375,39 +412,87 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
-		}
+	/* Setting test to active slot is not allowed. */
+	if (!confirm && slot == active_slot) {
+		return IMG_MGMT_ERR_INVALID_SLOT;
```
I think we should use proper errors for these, the slot isn't invalid, the current state disallows this

de-nordic replied:
Defined new error code.

nordicjm replied:
Still seems to be using invalid slot error

de-nordic replied:
Sorry, fixed wrong line. Now both lines are fixed and two new error codes added.

nordicjm reviewed (commented):
Error codes to update

nvlsianpu reviewed (dismissed):
my concerns were applied
